homeassistant/components/simplisafe/config_flow.py

```diff
@@ -97,17 +121,39 @@ async def _async_authenticate(
                         await asyncio.sleep(DEFAULT_EMAIL_2FA_SLEEP)
                     else:
                         break
-        except asyncio.TimeoutError:
-            return self.async_show_form(
-                step_id="user",
-                data_schema=STEP_USER_SCHEMA,
-                errors={"base": "2fa_timed_out"},
+        finally:
+            self.hass.async_create_task(
+                self.hass.config_entries.flow.async_configure(flow_id=self.flow_id)
+            )
+
+    async def async_step_email_2fa(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle email-based two-factor authentication."""
+        if not self._email_2fa_task:
+            self._email_2fa_task = self.hass.async_create_task(
+                self._async_get_email_2fa()
+            )
+            return self.async_show_progress(
+                step_id="email_2fa", progress_action="email_2fa"
             )
 
-        return await self._async_finish_setup()
+        try:
+            await self._email_2fa_task
```

Comment:
One curiosity I have: is there a way to detect if the user hits the `X` button in the UI? If so, I'd love to cancel this task, as there's no point in continuing a 5-minute run if the user gives up.

Reply:
I don't know that we have that kind of life cycle hook yet. I'd ask in Discord if someone knows more or has an idea.

Reply:
We currently don't have that, but we can add that by adding a hook here: https://github.com/home-assistant/core/blob/dev/homeassistant/data_entry_flow.py#L307